### PR TITLE
fix: Perform fields filtering server side

### PR DESCRIPTION
**Pull request by simster7**
Fixes: https://github.com/argoproj/argo/issues/4529

**Not ready for review**

Signed-off-by: Simon Behar <simbeh7@gmail.com>

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: util/json/json.go [hunk 1]
```diff
@@ -34,48 +34,3 @@ func (j *JSONMarshaler) NewEncoder(w io.Writer) gwruntime.Encoder {
 func (j *JSONMarshaler) Unmarshal(data []byte, v interface{}) error {
 	return json.Unmarshal(data, v)
 }
-
-// https://github.com/ksonnet/ksonnet/blob/master/pkg/kubecfg/diff.go
-func removeFields(config, live interface{}) interface{} {
```
Comment (simster7) [inline util/json/json.go:39]: This code is not used anywhere

File: go.mod [hunk 2]
```diff
@@ -19,6 +19,7 @@ require (
 	github.com/evanphx/json-patch v4.5.0+incompatible
 	github.com/fatih/structs v1.1.0 // indirect
 	github.com/gavv/httpexpect/v2 v2.0.3
+	github.com/ghodss/yaml v1.0.0
```
Comment (alexec) [inline go.mod:22]: remove?

File: util/fields/fields.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+			fieldsQuery = fieldsQuery[1:]
+			exclude = true
+		}
+		for _, field := range strings.Split(fieldsQuery, ",") {
+			fields[field] = true
+		}
+	}
+
+	data := make(map[string]interface{})
+	err := json.Unmarshal(dataBytes, &data)
```
[the remark this answers is not in the document]
Comment (simster7) [reply util/fields/fields.go:23]: > Prior-art - this is already how we do this. I.e. I'm wrong.

This is how we do this already. I did spend some modest time coming up with an alternative to no avail.

https://github.com/argoproj/pkg/blob/50e2680bec730dc985bb2e50d02c1d383caa0145/grpc/http/forwarders.go#L41-L108

File: go.mod (hunk 2 above)
Comment (simster7) [reply go.mod:22]: Yup

File: util/fields/fields.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexec) [reply util/fields/fields.go:23]: cool - thank you for trying

**Review by alexec — APPROVED** [review]
just remove the ghodds yaml and you're golden.

**Comment by alexec** [comment]
Merged into https://github.com/argoproj/argo/releases/tag/v2.12.0-rc4

**Pull request merged**